src/include/CacheManager.hpp

```diff
@@ -0,0 +1,56 @@
+/**   
+ * MIT License
+ *
+ * Copyright (c) 2021 Tsuhao Chang
+ * 
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef CACHE_MANAGER_H_
+#define CACHE_MANAGER_H_
+
+#include <string>
+#include <functional>
+#include <unordered_map>
+#include <mutex>
+
+#include "CacheContent.hpp"
+
+/**
+ * 將檔案內容載入記憶體
+ * 內部保存一個map，將key(應為request url)對應到一個CacheContent
+ */
+class CacheManager // singleton
+{
+public:
+    CacheManager();
+    ~CacheManager();
+    CacheManager &getInstance();
+    // 取得map內某個key的內容
+    // 如果該key不存在，建立一個新的CacheContent
+    // 如果該key存在但其CacheContent太舊，reload
+    // 如果該key存在但尚未ready，則其尚在讀檔，這段不確定該如何處理讓正在讀檔的worker知道讀檔完須多call一隻callback
+    // 如果該key存在且ready，回傳其內容
+    void get(std::string key, std::function<void(CacheContent &)> callback);
```

Comment:
const ref for input argument.

Reply:
Pass const CacheContent& into the callback if you don't want to allow the callback to modify the cache. Otherwise, pass in a pointer.

Reply:
fixed